### DISCUSSION TITLE
Optimize memory usage for LAMG and ConjugateGradient

### DISCUSSION THREAD

**Discussion by kisasexypantera94** [networkit/networkit#1292] [Q&A]
Currently ```LAMG``` and ```ConjugateGradient``` needs to store all RHSs and results at the same time, which is unnecessary. I suggest using lambda loaders and processors like so:
```c++
const auto rhs_loader = [&](size_t i)
{
    nk::Vector Y(m);
    ...
    return Y;
};

const auto result_processor = [&](size_t i, nk::Vector&& result)
{
    ...
};

size_t n_rhs = n;
size_t m_rhs = m;
lamg.parallelSolve(rhs_loader, result_processor, n_rhs, m_rhs);
```

if everything is ok, I will prepare a pull request

**Comment by avdgrinten** [comment]
What about passing a (mutable) reference to a nk::Vector to the RHS producer function? That way we could avoid memory allocations.

This would be a good addition, if we can re-implement the old API on top of the new one (and thus preserve backwards compatibility).

**Comment by kisasexypantera94** [comment]
Now that #765 is in the master, the question arises of how to use the new API in places where it is required to store all the solutions, as in ```CommuteTimeDistance```. Storing all solutions allows to quickly answer requests for the distance between any two nodes, but requires storing a huge (probably dense) matrix. In addition, the user may be interested in the distance between specific nodes. Maybe allow the user to pass a list of from-to node pairs as an argument and compute the distances for them in advance? Though the internals of the classes can get a little messy, there will be two arrays - ```solutions``` and ```distances```, the contents of which depend on the arguments.
I did not mention RHS vectors, they don't need to be stored anyway.
